packages/salesforce-adapter/src/filters/utils.ts

```diff
@@ -968,3 +975,96 @@ export const getProfilesAndPermissionSetsBrokenPaths = async (
   }
   return instance.value[PATHS_FIELD] ?? []
 }
+
+type CustomObjectsTargets = {
+  customObjects: readonly string[]
+  customObjectsLookups: Record<string, readonly string[]>
+}
+type SalesforceFetchTargets = {
+  metadataTypes: readonly string[]
+} & CustomObjectsTargets
+
+const isStringArray = (val: unknown): val is string[] => _.isArray(val) && val.every(_.isString)
+
+const isCustomObjectsTargets = (val: Values): val is CustomObjectsTargets =>
+  isStringArray(val[CUSTOM_OBJECTS_FIELD]) &&
+  _.isPlainObject(val[CUSTOM_OBJECTS_LOOKUPS_FIELD]) &&
+  Object.values(val[CUSTOM_OBJECTS_LOOKUPS_FIELD]).every(isStringArray)
+
+export const getOrgFetchTargets = async (elementsSource: ReadOnlyElementsSource): Promise<SalesforceFetchTargets> => {
+  const orgSettings = await elementsSource.get(FETCH_TARGETS_INSTANCE_ELEM_ID)
+  if (!isInstanceElement(orgSettings)) {
+    log.warn('Expected org settings Instance to be in elements source. Fetch targets will only include metadata types')
+    return {
+      metadataTypes: SUPPORTED_METADATA_TYPES,
+      customObjects: [],
+      customObjectsLookups: {},
+    }
+  }
+  const orgSettingsValues = orgSettings.value
+  if (!isCustomObjectsTargets(orgSettingsValues)) {
+    log.warn(
+      'Failed to extract customObject fetch settings from OrganizationSettings instance. Fetch targets will only include Metadata types',
+    )
+    log.trace('Singleton values are: %s', inspectValue(orgSettingsValues))
+    return {
+      metadataTypes: SUPPORTED_METADATA_TYPES,
+      customObjects: [],
+      customObjectsLookups: {},
+    }
+  }
+  return {
+    metadataTypes: SUPPORTED_METADATA_TYPES,
+    customObjects: orgSettingsValues[CUSTOM_OBJECTS_FIELD],
+    customObjectsLookups: orgSettingsValues[CUSTOM_OBJECTS_LOOKUPS_FIELD],
+  }
+}
+
+const getCustomObjectDependenciesRecursively = (
+  target: string,
+  lookups: Record<string, readonly string[]>,
+  handledTypes: Set<string>,
+): string[] => {
+  if (handledTypes.has(target)) {
+    return []
+  }
+  handledTypes.add(target)
+  const dependencies = lookups[target] ?? []
+  return _.uniq(
+    dependencies.reduce(
+      (acc, dep) => acc.concat(dep, getCustomObjectDependenciesRecursively(dep, lookups, handledTypes)),
+      [target],
+    ),
+  )
+}
+
+export const getMetadataIncludeFromFetchTargets = async (
```

Comment:
Nit: This function is also for data.
```suggestion
export const getMetadataQueryIncludeFromFetchTargets = async (
```

Reply:
Fetch targets are referring to the include section within `Metadata` though 🤔